### DISCUSSION TITLE
What is the status of the project?

### DISCUSSION THREAD

**Discussion by nbelyh** [maxGraph/maxGraph#64] [Announces]
Dear all, I am a newcomer, interested in keeping mxgraph alive as a library and not in a form of "diagrams.net"
Could you kindly update me, what is the status of this fork - is somebody working on this? 
It looks like there is not much happening. Is it a dead project?

**Comment by tbouffard** [comment]
Hello @nbelyh 
The project is alive, work is done in the `development` branch.
For now, we are looking for people ready to build and test the refactored code. See #51 for more details.
About a future release, see #41

**Reply by shakeZSL** [reply]
Hello everybody,
i hope its okay to ask like this, i am still a student and unsure about the right procedure.

I am trying to evaluate a diagram library for my company. At the moment we are unsure if we should use mxGraph because it is 
end of life. 
However your fork and especially your TypeScript implementation are looking very promising.
I tried to test maxGraph with the 2 given commands: "npm install", "npm run dev".
If i build the core it works, however when i chose "maxGraph/html" i get the following error:
`Running 'npx lerna run dev --stream --scope @maxgraph/html'...
lerna notice cli v4.0.0
lerna notice filter including "@maxgraph/html"
lerna info filter [ '@maxgraph/html' ]
lerna info Executing command in 1 package: "npm run dev"
@maxgraph/html: > @maxgraph/html@1.0.0 dev D:\maxGraph\maxGraph-development\packages\html
@maxgraph/html: > cross-env NODE_OPTIONS=--openssl-legacy-provider start-storybook -p 8901 -s ./public
@maxgraph/html: node: --openssl-legacy-provider is not allowed in NODE_OPTIONS
@maxgraph/html: npm ERR! code ELIFECYCLE
@maxgraph/html: npm ERR! errno 9
@maxgraph/html: npm ERR! @maxgraph/html@1.0.0 dev: `cross-env NODE_OPTIONS=--openssl-legacy-provider start-storybook -p 8901 -s ./public`    
@maxgraph/html: npm ERR! Exit status 9
@maxgraph/html: npm ERR!
@maxgraph/html: npm ERR! Failed at the @maxgraph/html@1.0.0 dev script.`

I tried integrating the project in different ways but couldnt get it to work either.
I guess the above error is a mistake on my part, if so could you guide me how to build the project in the right way? 
Also do you have a rough idea, when you will be able to provide a complete npm package?
I would be happy to test the library and report bugs i find along the way.

Thank you for your time!

**Reply by tbouffard** [reply]
hi @nbelyh @shakeZSL 
I have the same issue, the repo must be fixed.
In the meantime, you can apply this workaround: in the `packages/html/package.json` file, remove the `cross-env NODE_OPTIONS=--openssl-legacy-provider` part, the server will start.
I have tested this with node v16, npm v8

Next time, please open a dedicated discussion to keep threads focused on a single topic.

**Reply by mcyph** [reply]
Yeah, I added that line to make the storybook examples run on node 17, but it unfortunately it looks like it breaks earlier versions. `--openssl-legacy-provider` is a workaround anyway, so will need to investigate whether there's a better solution. See also https://github.com/storybookjs/storybook/issues/16555

**Reply by nbelyh** [reply]
> @tbouffard 
> hi @nbelyh

Probably the right person to CC: @shakeZSL 
I think I will be stuck for now with "classic" mxgraph, until this fork is more stable. Probably will look at it again in a couple of month/years 😅

**Reply by mcyph** [reply]
@nbelyh I don't think we're hugely far off getting a release once getting the serialization and possibly the styling system worked out, but until then I think that's the best choice, as there are still breaking changes to be made. 

The examples and explanations need to be updated within the comments. In some ways the new docs are more complete/correct than the old ones in terms of the required types etc, and most examples have already been ported to storybook (aside from the editor). Will try to work on the remaining ones as soon as feasible.

**Reply by shakeZSL** [reply]
Thank you for your suggestion, the fix worked great!

I guess we will follow  @nbelyh's  advice. However i will still try to test some stuff.
If i find anything i will create an issue.
@tbouffard i will create a dedicated discussion next time, thanks :)

**Reply by tbouffard** [reply]
I have just created  #74 to fix the storybook startup error

**Reply by tbouffard** [reply]
this has been merged!

**Comment by mcyph** [comment]
It's been on a a hiatus for a while, but @junsikshim made some pretty major progress a few months ago, and I've made reasonably large changes in the last few days in my branch at https://github.com/mcyph/mxGraph/tree/development.

A big problem is that on the one hand many existing applications rely on the current version. On the other, it would be very tricky to make more than incremental changes to e.g. add support for new browser versions because of how complex and tightly integrated the library is. This is from the original authors in reference to the draw.io/diagrams.net project, but I think a lot of this also applies to mxGraph (maxGraph is an unofficial fork which started after development on mxGraph stopped):

> Similar to SQLite, diagrams.net is open source but closed to contributions.
>
> The level of complexity of this project means that even simple changes can break a lot of other moving parts. The amount of testing required is far more than it first seems. If we were to receive a PR, we'd have to basically throw it away and write it how we want it to be implemented.
>
> We are grateful for community involvement, bug reports, & feature requests. We do not wish to come off as anything but welcoming, however, we've made the decision to keep this project closed to contributions for the long term viability of the project.

Just like the touch element event issues you ran into, much of the logic of maxGraph extends across many files, and making one change can have unintended consequences. My hope was by introducing TypeScript and recent JavaScript syntax that it could reduce the cascade effect and allow more people to understand and build on the code. 

Unfortunately, in trying to sort out e.g. global assigning of variables to prototypes and the state of one graph affecting another it was pretty much inevitable that compatibility would be broken to some degree, potentially needing a rewrite of parts of previously written applications, if they would be feasible. For this reason, I think there is a case to keep both the existing version and the rewrite maintained, but personally for both real life reasons (juggling uni/work/other projects/PC instability..), and practical ones like that I don't have much experience with the previous version of the library mean that I could only provide minimal support for the older version, and so have mainly been focusing my efforts on the new code base.

There are currently no tests, so whether things function correctly relies on manual testing - I think this may be inevitable because writing unit tests for something like this would be a gigantic task, if even possible for much of it. The currently available TypeScript definition files were a good first step, but they along with the types in the original documentation were incomplete and contained many inaccuracies which would be difficult to solve without coding the library itself in TypeScript.

This is the first time I've had a chance in the last few months to develop, and am trying to focus on the following areas in the last few days:

- Make the directory structure simpler and more logical
- Convert to TypeScript every JavaScript file remaining in the codebase
- Eliminate all the underlined TypeScript errors
- Create up-to-date documentation:
  - Auto-generated API docs from the JSDoc in each file using something like TypeDoc (or similar)
  - Provide a tutorial for new users and porting guide
  - Provide an intro to the codebase for contributing

The first couple of items are already some way towards being completed. I'd also like to finish off all the examples, including some ones which haven't yet been updated for the changed code, but think will work on the specifics after working thru the above issues.

**Reply by tbouffard** [reply]
Great news @mcyph , I will also have time to work on maxGraph next week.
I plan to do some manual tests (AFAIK mxGraph had no automatic tests, and a lot of regressions appeared in the past - the past maintainers was releasing internal alpha/beta mxGraph versions on daily basis, you can see that in the draw.io commit/change log), mainly on Firefox (I have already detected issues) and write some documentation, both about the API and the migrating/porting guide. I know some parts of the mxGraph old implementation, both code and API. I will contact you when I am available for work.

In the meantime, I suggest we prepare a alpha release of maxGraph to let people easily try the lib as a regular npm dependency (no need for building the lib).

**Reply by nbelyh** [reply]
Are you planning to get rid of IE6 specific stuff? Like [VML](https://en.wikipedia.org/wiki/Vector_Markup_Language) renderer and [quirks mode](https://en.wikipedia.org/wiki/Quirks_mode) for example. Before refactoring, probably it could make sense to get rid of some old code that is hardly to be ever used again, so it is less code to refactor? I believe by removing these two alone the code base could be reduced by 10-20%.

According to my experience, the strategy "let's rewrite it all from scratch in language x" never works for big projects. The passion is gone after a couple of months, and that's where it all ends. I would probably start with providing some automatic build (first) that produces an NPM package (so that you have a notion of "release"), and then some sort of tests so that one can be somewhat sure that the changes are not breaking, and removing code from year 2000. Also writing tests help to understand the code much better.

**Reply by junsikshim** [reply]
> "let's rewrite it all from scratch in language x"

We'll never do that :)
We've already removed all the VML stuff from the codebase as well as some other 'old' stuff.
As @mcyph mentioned, it's near impossible to write tests for all the functionalities, so instead we brought in Storybook for the examples to make the manual testing convenient.

While refactoring, some breaking changes have been made(including the parameters of Graph(main) class). So it may bring some unintended side-effects, but with the examples updated accordingly, things are working pretty smoothly. You can check out `development` branch before a release is made.

I'm still very busy with work, so I'll be back in January (I hope).

**Reply by mcyph** [reply]
Yeah exactly, there's no way I'd even think of rewriting from scratch. This might change in the future in terms of e.g. new shapes or similar but I wouldn't even have plans for significantly changing the functionality at this stage - it's mainly refactoring and documentation so we can understand it better. I've refactored similarly-sized codebases into TypeScript before, though it is quite an energy-intensive process and takes time. I think IE10 onwards could be reasonable to support in the new version - for older applications which aren't being updated maybe/maybe it doesn't make sense to keep support for earlier versions of IE with the Trident engine. IE6 has well and truly gone into sunset phase by MS, and to be honest I'm not sure it's worth spending time and resources supporting such an old and insecure version.

I don't know exactly how the original version of jGraph was developed, but looking at the Java codebase there's a lot of similar layout and architecture to the JS version, even if it has diverged in the way it was implemented. I suspect that if it was originally written in a strongly-typed language the original devs may have been able to use it as reference for writing the JS version and that that could've eased transferring some of the layout and functionality to a weakly-typed prototype-based language like JS.

Even though I worked on the codebase previously it took me some time to get up to speed and initially didn't know what would be involved with finishing off the annotations etc. Things have come a long way I think mostly for the better in the last few months, and after reorganising the directory structure (which I ironically put in place to make identifying common functionality reducing files with say 10k+ lines of code into smaller ones) into one which was more easy to grasp it seems a lot more feasible to me now. 

Up till now there's been a focus on getting things done given limited time, but I'd like to focus on documenting+standardising the structure of the codebase to allow new developers to be able to grasp things more quickly and easily in the future. While it might be true that 500kb zipped isn't that much nowadays, it still is pretty slow to load that amount of (unzipped) code on say underpowered mobile phones etc, and if done well I think the separation of concerns into smaller pieces which do more targeted things could make the codebase significantly easier to understand and build on.

**Reply by mcyph** [reply]
@junsikshim btw would it be OK if I moved the `core` and `html` directories inside `packages/` into the root maxGraph directory? I've been thinking it could be a good idea to reduce the number of subdirectories, but didn't want to do it in case you had a reason for organising in this way. Not sure what to do about the `react` subdir too, thinking it may be a good idea to delete that for now, focus on getting an mvp released and add support later

**Reply by nbelyh** [reply]
@junsikshim Thank you for the detailed response. So, VML is already gone? As for me, I don't need any IE (Trident) at all. I doubt anyone else does, looking into the future. But it would be great to have something releasable, i.e. not a "set of files I am playing with when I have free time" but something like the normal releases with a CI system backing it up 😄 Something users like me could use in their project without including full source, I'm thinking an NPM module that is built by GitHub actions and hosted by NPMJS or something like this.

**Reply by mcyph** [reply]
@nbelyh yep, VML is already gone, removed many months ago - I accidently removed some mixed HTML stuff I shouldn't have before I understood what it did though and want to add it back as soon as possible..I do fully agree with you re the "set of files I am playing with when I have free time" bit but my experience is the time between "there's so much to annotate and get done" and "it's all annotated and not much more to do" is surprisingly short+it takes some time to get there. Once it gets to that stage I think that will be much more feasible.

**Reply by mcyph** [reply]
still have some way to go before being able to merge with development, but have created a new branch at https://github.com/maxGraph/maxGraph/tree/bleeding-edge with the recent changes

**Reply by junsikshim** [reply]
@mcyph Having `packages/` at the root directory is an indication of a monorepo, so I think leaving the structure as-is can help the newcomers to understand the project easier. Also, without `packages/`, `core` and `html` will exist along side of `docs` and `scripts` which are not really related to the source code.

I definitely agree with removing the `react` directory. It was meant to show there can be such package for each framework, but I don't think we'll be able to create the react examples anytime soon :)

**Reply by mcyph** [reply]
@tbouffard @junsikshim @nbelyh I've finished off a majority of the red underlined errors and got most the examples in my bleeding-edge branch working again. There's a small amount of bugs which were either in the original version mxGraph and identified by TypeScript which still need to be fixed and a few files left, but all but a handful of the files have been annotated with types, which means a lot of the changes which were previously tricky should hopefully be much easier from now onwards. Hoping can stabilise this soon+finish off the annotations so that we can use TypeDoc or similar to generate the API documentation. Updating the API docs and finishing off the unconverted/broken examples will be a large amount of what still needs to be done, but we've already done a reasonable amount of the groundwork and will update with more details soon!

**Reply by mcyph** [reply]
Finished off all annotations+TypeDoc generation now works (a preview is at https://mcyph.github.io/mxgraph/api/). This is about all I can do for now - I'll write some docs on what's changed and migration etc soon. The parameters and types themselves should mostly be complete, and I think they should be useful as-is, but example code and explanations need to be updated.

**Comment by bsanders1979** [comment]
Hello. I just want to confirm the status of this project. It looks active, but I feel like I'm seeing conflicting information, so I just want to know if this is something that could be used for a project that will start small with potential for growth. Thanks.

**Reply by tbouffard** [reply]
Hi yes the project is alive. The Readme has been updated recently to provide more accurate information.
We are currently working on the license #89 and versioning #92.
Then we will be able to release a first alpha version and publish a package on npmjs.

**Comment by sadek-hussein** [comment]


**Reply by tbouffard** [reply]
Please post in a dedicated topic and remove this comment

**Reply by sadek-hussein** [reply]
done

**Comment by tbouffard** [comment]
The first release of `maxGraph`  is now available: see #147.

**Reply by csouchet** [reply]
🎉 🎉 🎉
